File: web/src/permalink.test.ts
// Locks the canonical session permalink format (F3). agentd mints the same
// string server-side (go/cmd/agentd/permalink.go) — if this table changes, that
// one changes with it.

import { describe, it, expect } from 'vitest'
import {
  SESSION_PERMALINK_FORMAT,
  buildSessionPath,
  buildSessionPermalink,
  parseSessionPermalink,
} from './permalink.js'

describe('buildSessionPath', () => {
  const cases: Array<{ name: string; project: string; session: string; want: string }> = [
    { name: 'plain ids', project: 'acme', session: 'sess-123', want: '/p/acme/s/sess-123' },
    { name: 'uuid session', project: 'p1', session: '9f1c2d3e-0000-4a5b-8c7d-000000000001',
      want: '/p/p1/s/9f1c2d3e-0000-4a5b-8c7d-000000000001' },
    { name: 'encodes slashes', project: 'a/b', session: 'c d', want: '/p/a%2Fb/s/c%20d' },
    { name: 'encodes query chars', project: 'p?x', session: 's#y', want: '/p/p%3Fx/s/s%23y' },
  ]
  for (const c of cases) {
    it(c.name, () => {
      expect(buildSessionPath(c.project, c.session)).toBe(c.want)
    })
  }
})

describe('buildSessionPermalink', () => {
  const cases: Array<{ name: string; base: string; want: string }> = [
    { name: 'empty base → relative', base: '', want: '/p/acme/s/s1' },
    { name: 'origin', base: 'https://orange.example.com', want: 'https://orange.example.com/p/acme/s/s1' },
    { name: 'trailing slash trimmed', base: 'https://orange.example.com/', want: 'https://orange.example.com/p/acme/s/s1' },
    { name: 'many trailing slashes trimmed', base: 'https://o.example.com///', want: 'https://o.example.com/p/acme/s/s1' },
    { name: 'sub-path base preserved', base: 'https://o.example.com/agents', want: 'https://o.example.com/agents/p/acme/s/s1' },
    { name: 'localhost with port', base: 'http://localhost:8080', want: 'http://localhost:8080/p/acme/s/s1' },
  ]
  for (const c of cases) {
    it(c.name, () => {
      expect(buildSessionPermalink(c.base, 'acme', 's1')).toBe(c.want)
    })
  }
})

describe('parseSessionPermalink', () => {
  const ok: Array<{ name: string; input: string; project: string; session: string }> = [
    { name: 'relative path', input: '/p/acme/s/s1', project: 'acme', session: 's1' },
    { name: 'absolute url', input: 'https://o.example.com/p/acme/s/s1', project: 'acme', session: 's1' },
    { name: 'http url with port', input: 'http://localhost:8080/p/acme/s/s1', project: 'acme', session: 's1' },
    { name: 'sub-path mount', input: 'https://o.example.com/agents/p/acme/s/s1', project: 'acme', session: 's1' },
    { name: 'trailing slash', input: '/p/acme/s/s1/', project: 'acme', session: 's1' },
    { name: 'query string ignored', input: '/p/acme/s/s1?tab=artifacts', project: 'acme', session: 's1' },
    { name: 'hash ignored', input: '/p/acme/s/s1#msg-4', project: 'acme', session: 's1' },
    { name: 'hash before query', input: '/p/acme/s/s1#a?b', project: 'acme', session: 's1' },
    { name: 'percent-decoded', input: '/p/a%2Fb/s/c%20d', project: 'a/b', session: 'c d' },
  ]
  for (const c of ok) {
    it(`parses ${c.name}`, () => {
      expect(parseSessionPermalink(c.input)).toEqual({ projectId: c.project, sessionId: c.session })
    })
  }

  const bad: Array<{ name: string; input: string }> = [
    { name: 'empty', input: '' },
    { name: 'root', input: '/' },
    { name: 'project only', input: '/p/acme' },
    { name: 'wrong session segment', input: '/p/acme/x/s1' },
    { name: 'wrong project segment', input: '/q/acme/s/s1' },
    { name: 'missing session id', input: '/p/acme/s/' },
    { name: 'unrelated path', input: '/settings/workers' },
    { name: 'origin only', input: 'https://o.example.com' },
  ]
  for (const c of bad) {
    it(`rejects ${c.name}`, () => {
      expect(parseSessionPermalink(c.input)).toBeNull()
    })
  }

  it('round-trips every build output', () => {
    const ids: Array<[string, string]> = [
      ['acme', 's1'],
      ['a/b', 'c d'],
      ['proj-with-dashes', '9f1c2d3e-0000-4a5b-8c7d-000000000001'],
      ['ünï', 'sé§sion'],
    ]
    for (const [project, session] of ids) {
      const url = buildSessionPermalink('https://o.example.com', project, session)
      expect(parseSessionPermalink(url)).toEqual({ projectId: project, sessionId: session })
    }
  })

  it('malformed percent-escapes do not throw', () => {
    expect(parseSessionPermalink('/p/%E0%A4%A/s/s1')).toEqual({
      projectId: '%E0%A4%A',
      sessionId: 's1',
    })
  })
})

it('documents the canonical format', () => {
  expect(SESSION_PERMALINK_FORMAT).toBe('/p/:projectId/s/:sessionId')
})

File: go/cmd/agentd/permalink_test.go
package main

import "testing"

func TestResolvePublicBaseURL(t *testing.T) {
	cases := []struct {
		name string
		env  map[string]string
		want string
	}{
		{"default", nil, "http://localhost:8080"},
		{"empty value falls back to default", map[string]string{"AGENTKIT_PUBLIC_BASE_URL": ""}, "http://localhost:8080"},
		{"https origin", map[string]string{"AGENTKIT_PUBLIC_BASE_URL": "https://orange.example.com"}, "https://orange.example.com"},
		{"trailing slash trimmed", map[string]string{"AGENTKIT_PUBLIC_BASE_URL": "https://orange.example.com/"}, "https://orange.example.com"},
		{"many trailing slashes trimmed", map[string]string{"AGENTKIT_PUBLIC_BASE_URL": "https://orange.example.com///"}, "https://orange.example.com"},
		{"surrounding space trimmed", map[string]string{"AGENTKIT_PUBLIC_BASE_URL": "  https://orange.example.com  "}, "https://orange.example.com"},
		{"sub-path preserved", map[string]string{"AGENTKIT_PUBLIC_BASE_URL": "https://example.com/agents"}, "https://example.com/agents"},
		{"port preserved", map[string]string{"AGENTKIT_PUBLIC_BASE_URL": "http://10.0.0.4:8080"}, "http://10.0.0.4:8080"},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			p, err := resolvePublicBaseURL(envMap(tc.env))
			if err != nil {
				t.Fatal(err)
			}
			if p.BaseURL() != tc.want {
				t.Fatalf("BaseURL() = %q, want %q", p.BaseURL(), tc.want)
			}
		})
	}
}

func TestResolvePublicBaseURL_Invalid(t *testing.T) {
	cases := []struct {
		name  string
		value string
	}{
		{"no scheme", "orange.example.com"},
		{"relative path", "/agents"},
		{"non-http scheme", "ftp://orange.example.com"},
		{"scheme only", "https://"},
		{"carries query", "https://orange.example.com?x=1"},
		{"carries fragment", "https://orange.example.com#top"},
		{"only slashes", "///"},
		{"unparseable", "http://[::1"},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if _, err := resolvePublicBaseURL(envMap(map[string]string{
				"AGENTKIT_PUBLIC_BASE_URL": tc.value,
			})); err == nil {
				t.Fatalf("expected error for %q", tc.value)
			}
		})
	}
}

// The format is the contract with web/src/permalink.ts — these expectations
// mirror its permalink.test.ts table.
func TestPermalinkerSessionURL(t *testing.T) {
	cases := []struct {
		name    string
		base    string
		project string
		session string
		want    string
	}{
		{"default base", "", "acme", "sess-1", "http://localhost:8080/p/acme/s/sess-1"},
		{"custom origin", "https://orange.example.com", "acme", "sess-1", "https://orange.example.com/p/acme/s/sess-1"},
		{"sub-path mount", "https://example.com/agents", "acme", "sess-1", "https://example.com/agents/p/acme/s/sess-1"},
		{"uuid session", "https://o.example.com", "p1", "9f1c2d3e-0000-4a5b-8c7d-000000000001",
			"https://o.example.com/p/p1/s/9f1c2d3e-0000-4a5b-8c7d-000000000001"},
		{"escapes slashes", "https://o.example.com", "a/b", "c/d", "https://o.example.com/p/a%2Fb/s/c%2Fd"},
		{"escapes spaces", "https://o.example.com", "a b", "c d", "https://o.example.com/p/a%20b/s/c%20d"},
		{"escapes query chars", "https://o.example.com", "p?x", "s#y", "https://o.example.com/p/p%3Fx/s/s%23y"},
		{"empty session id → no link", "https://o.example.com", "acme", "", ""},
		{"empty project id → no link", "https://o.example.com", "", "sess-1", ""},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			env := map[string]string{}
			if tc.base != "" {
				env["AGENTKIT_PUBLIC_BASE_URL"] = tc.base
			}
			p, err := resolvePublicBaseURL(envMap(env))
			if err != nil {
				t.Fatal(err)
			}
			if got := p.SessionURL(tc.project, tc.session); got != tc.want {
				t.Fatalf("SessionURL(%q, %q) = %q, want %q", tc.project, tc.session, got, tc.want)
			}
		})
	}
}
